### PR TITLE
Move mobile app QR code to Grafana user profile page

### PR DESCRIPTION
**Pull request by joeyorlando**
# What this PR does

Dependent on https://github.com/grafana/grafana/pull/77863

## Checklist

- [ ] Unit, integration, and e2e (if applicable) tests updated
- [ ] Documentation added (or `pr:no public docs` PR label added if not required)
- [ ] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)

### CODE REVIEW

File: grafana-plugin/src/module.ts [hunk 1]
```diff
@@ -1,15 +1,49 @@
...
+  plugin.configureExtensionLink({
+    title: 'OnCall stuff!',
+    description: 'hello world',
+    extensionPointId: PluginExtensionPoints.ExploreToolbarAction,
+    category: 'OnCall',
+    icon: 'fire',
+    onClick: async () => {
+      console.log('FROM WITHIN ONCALL, HI!');
+    },
+  });
```
Comment (joeyorlando) [inline grafana-plugin/src/module.ts:46]: neither of these approaches seem to work (yet) as I would expect

File: grafana-plugin/src/module.ts [hunk 2]
```diff
@@ -1,15 +1,35 @@
...
   body: PluginConfigPage as unknown as ComponentClass<OnCallPluginConfigPageProps, unknown>,
   id: 'configuration',
 });
+
+if (
+  'configureExtensionLink' in plugin &&
+  PluginExtensionPoints != null &&
+  'ExploreToolbarAction' in PluginExtensionPoints
+) {
+  plugin.configureExtensionComponent({
```
Comment (joeyorlando) [inline grafana-plugin/src/module.ts:23]: here is [the only example I could find](https://github.com/grafana/grafana-pdc-app/blob/e39bd2f2ca6cf5cf1c11cd96805f2dfd6b08b505/src/module.ts#L37) of another plugin using `configureExtensionComponent`

File: grafana-plugin/src/module.ts
Comment (joeyorlando) [inline grafana-plugin/src/module.ts:1]: see [here](https://github.com/grafana/incident/blob/main/packages/grafana-incident-app/src/module.tsx#L103-L128) for an example of how Grafana Incident is registering extension points

File: grafana-plugin/src/plugin.json [hunk 3]
```diff
@@ -3,6 +3,7 @@
   "type": "app",
   "name": "Grafana OnCall",
   "id": "grafana-oncall-app",
+  "preload": true,
```
Comment (joeyorlando) [inline grafana-plugin/src/plugin.json:6]: without this Grafana wasn't properly registering our extension points. See the discussion in Slack [here](https://raintank-corp.slack.com/archives/C01C4K8DETW/p1699444399748929?thread_ts=1699444056.495409&cid=C01C4K8DETW) for more context

File: grafana-plugin/src/module.ts
Comment (joeyorlando) [reply grafana-plugin/src/module.ts:1]: and [here](https://grafana.com/developers/plugin-tools/ui-extensions/register-an-extension) are the Grafana Plugin Tools docs on Extension Points (note that `configureExtensionComponent` is currently not documented; see [this](https://raintank-corp.slack.com/archives/C064R17Q1A8/p1699375441201929) Slack thread for a conversation with the plugins platform team).

File: docker-compose-developer.yml [hunk 4]
```diff
@@ -52,7 +52,7 @@ services:
       dockerfile: Dockerfile.dev
       labels: *oncall-labels
     environment:
-      ONCALL_API_URL: http://host.docker.internal:8080
+      ONCALL_API_URL: http://localhost:8080
```
Comment (joeyorlando) [inline docker-compose-developer.yml:55]: NOTE: remove this change before merging

File: grafana-plugin/src/components/ExtensionLinkMenu/ExtensionLinkMenu.tsx [hunk 5]
```diff
@@ -69,7 +67,7 @@ function DeclareIncidentMenuItem({ extensions, declareIncidentLink, grafanaIncid
             category: 'Incident',
             title: 'Declare incident',
             pluginId: 'grafana-oncall-app',
-          },
+          } as any,
```
Comment (joeyorlando) [inline grafana-plugin/src/components/ExtensionLinkMenu/ExtensionLinkMenu.tsx:70]: is there an alternative to using `any` here?

File: grafana-plugin/src/containers/MobileAppConnection/MobileAppConnection.tsx [hunk 6]
```diff
@@ -23,7 +23,8 @@ import QRCode from './parts/QRCode/QRCode';
 const cx = cn.bind(styles);
 
 type Props = {
-  userPk: User['pk'];
+  userPk?: User['pk'];
+  // store: typeof rootStore;
```
Comment (joeyorlando) [inline grafana-plugin/src/containers/MobileAppConnection/MobileAppConnection.tsx:27]: is `store` not needed here anymore?

File: grafana-plugin/src/module.ts [hunk 7]
```diff
@@ -1,15 +1,47 @@
...
+const plugin = new AppPlugin<OnCallPluginMetaJSONData>().setRootPage(GrafanaPluginRootPage).addConfigPage({
   title: 'Configuration',
   icon: 'cog',
   body: PluginConfigPage as unknown as ComponentClass<OnCallPluginConfigPageProps, unknown>,
   id: 'configuration',
 });
+
+if (isUseProfileExtensionPointEnabled()) {
+  // TODO Change this to use the PluginExtensionPoints enum once that is available in @grafana/data (>10.2.2)
+  const extensionPointId = 'grafana/user/profile/tab';
```
Comment (joeyorlando) [inline grafana-plugin/src/module.ts:21]: this can _probably_ be updated to `PluginExtensionPoints.UserProfileTab `

File: grafana-plugin/src/utils/index.ts [hunk 8]
```diff
@@ -95,3 +96,10 @@ export function getPaths(obj?: any, parentKey?: string): string[] {
 export function pluralize(word: string, count: number): string {
   return count === 1 ? word : `${word}s`;
 }
+
+export function isUseProfileExtensionPointEnabled(): boolean {
+  const { major, minor } = getGrafanaVersion();
+  const isRequiredGrafanaVersion = major > 10 || (major === 10 && minor >= 3);
+
+  return isRequiredGrafanaVersion;
+}
```
Comment (joeyorlando) [inline grafana-plugin/src/utils/index.ts:105]: this seems duplicate of [this function](https://github.com/grafana/oncall/pull/3296/files#diff-bc9e12d72092a3f2944bb5bb941ee47f12f58c37e4dcdfbfd667cac49b83876bR35-R45) defined in `module.ts`

File: grafana-plugin/src/containers/UserSettings/parts/index.tsx [hunk 9]
```diff
@@ -140,7 +141,9 @@ export const TabsContent = observer(({ id, activeTab, onTabChange, isDesktopOrLa
         ) : (
           <PhoneVerification userPk={id} />
         ))}
-      {activeTab === UserSettingsTab.MobileAppConnection && <MobileAppConnection userPk={id} />}
+      {!isUseProfileExtensionPointEnabled() && activeTab === UserSettingsTab.MobileAppConnection && (
+        <MobileAppConnection userPk={id} />
+      )}
```
Comment (joeyorlando) [inline grafana-plugin/src/containers/UserSettings/parts/index.tsx:146]: @Matvey-Kuk brought up a good point that we _might_ still want to show this tab on Grafana >= 10.3, but instead of showing the QR code content, show a message w/ a link pointing users to the new tab on the `/profile` page (you should be able to deep link to the "IRM" tab using `/profile?tab=irm`)

File: grafana-plugin/src/utils/index.ts (hunk 8 above)
Comment (teodosii) [reply grafana-plugin/src/utils/index.ts:105]: Yea this one is actually tricky because it adds a loophole if you reference stuff in module.ts from my attempts

File: grafana-plugin/src/module.ts (hunk 7 above)
Comment (teodosii) [reply grafana-plugin/src/module.ts:21]: That was not on the PluginExtensionPoints last time I checked, I believe that package doesn't have tge change yet

File: grafana-plugin/src/containers/MobileAppConnection/MobileAppConnection.tsx (hunk 6 above)
Comment (teodosii) [reply grafana-plugin/src/containers/MobileAppConnection/MobileAppConnection.tsx:27]: hmm was it ever needed?

Comment (teodosii) [reply grafana-plugin/src/containers/MobileAppConnection/MobileAppConnection.tsx:27]: it's just the comment, needs to be removed

File: grafana-plugin/src/containers/UserSettings/parts/connectors/Connectors.tsx [hunk 10]
```diff
@@ -8,6 +8,7 @@ import { AppFeature } from 'state/features';
 import { useStore } from 'state/useStore';
 
 import ICalConnector from './ICalConnector';
+// import MobileAppConnector from './MobileAppConnector';
```
Comment (brojd) [inline grafana-plugin/src/containers/UserSettings/parts/connectors/Connectors.tsx:11]: can be removed?

File: grafana-plugin/src/plugin/GrafanaPluginRootPage.helpers.tsx [hunk 11]
```diff
@@ -4,6 +4,21 @@ export function isTopNavbar(): boolean {
   return !!config.featureToggles.topnav;
 }
 
+export function getGrafanaVersion(): { major?: number; minor?: number; patch?: number } {
```
Comment (brojd) [inline grafana-plugin/src/plugin/GrafanaPluginRootPage.helpers.tsx:7]: can we unit test this one?

File: grafana-plugin/src/utils/index.ts (hunk 8 above)
Comment (brojd) [reply grafana-plugin/src/utils/index.ts:105]: Maybe extracting that to separate file and importing it from both `module.ts` and this file will help with circular dep?

**Review by brojd — APPROVED** [review]
some minor suggestions but looks good in general

File: grafana-plugin/src/utils/index.ts (hunk 8 above)
Comment (teodosii) [reply grafana-plugin/src/utils/index.ts:105]: You still need to reference the module file back because there's the plugin defined - for this line `'configureExtensionComponent' in plugin`

File: grafana-plugin/e2e-tests/globalSetup.ts [hunk 12]
```diff
@@ -21,6 +28,13 @@ import { goToGrafanaPage } from './utils/navigation';
 
 const grafanaApiClient = new GrafanaAPIClient(GRAFANA_ADMIN_USERNAME, GRAFANA_ADMIN_PASSWORD);
 
+enum OrgRole {
+  None = 'None',
+  Viewer = 'Viewer',
+  Editor = 'Editor',
+  Admin = 'Admin',
+}
```
Comment (joeyorlando) [inline grafana-plugin/e2e-tests/globalSetup.ts:36]: what is this change related to?

File: grafana-plugin/src/module.ts [hunk 13]
```diff
@@ -1,15 +1,48 @@
...
+
+  plugin.configureExtensionComponent({
+    title: IRM_TAB,
+    description: 'IRM settings',
+    extensionPointId,
+    /**
+     * typing MobileAppConnectionWrapper as any until 10.2.0 is released
+     * https://github.com/grafana/grafana/pull/75019#issuecomment-1724997540
+     */
```
Comment (joeyorlando) [inline grafana-plugin/src/module.ts:31]: I believe I copied this comment from another repo where I found how they were doing this.. this comment might be irrelevant. You could try removing the `as any` and see if ts complains or not?

File: grafana-plugin/e2e-tests/globalSetup.ts (hunk 12 above)
Comment (teodosii) [reply grafana-plugin/e2e-tests/globalSetup.ts:36]: We don't mark the package.json to be a module, while grafana I believe it does, which leads to conflicts when importing it. Playwright will complain on the import from `grafana/data` of `OrgRole`, you'll get this error

`Error [ERR_REQUIRE_ESM]: require() of ES Module not supported [duplicate]`

Which would sort of be fixed by adding this to the `package.json`

`"type": "module"`

but this means we're only allow to use ES6 imports, therefore no `require` imports anymore, and I believe that would be conflicting with the libraries we depend on that use `requrie` in their code.

I'm not an expert on this aspects, so this might need additional research, for now I've removed the import and just set the enum in our codebase instead, couldn't figure out a better way instead :thinking: Changing the codebase to `module` leads to other problems

File: grafana-plugin/src/module.ts (hunk 13 above)
Comment (teodosii) [reply grafana-plugin/src/module.ts:30]: Right, not needed.

File: grafana-plugin/src/plugin/GrafanaPluginRootPage.helpers.tsx (hunk 11 above)
Comment (teodosii) [reply grafana-plugin/src/plugin/GrafanaPluginRootPage.helpers.tsx:7]: Added the unit test

**Review by vadimkerr — APPROVED** [review]

File: grafana-plugin/src/module.ts (hunk 7 above)
Comment (teodosii) [reply grafana-plugin/src/module.ts]: done now since it's available in the latest package

File: grafana-plugin/src/components/ExtensionLinkMenu/ExtensionLinkMenu.tsx (hunk 5 above)
Comment (teodosii) [reply grafana-plugin/src/components/ExtensionLinkMenu/ExtensionLinkMenu.tsx]: done
